cpp/src/io/text/multibyte_split.cu

```diff
@@ -247,18 +308,134 @@ std::vector<rmm::cuda_stream_view> get_streams(int32_t count, rmm::cuda_stream_p
   return streams;
 }
 
-int64_t multibyte_split_scan_full_source(cudf::io::text::data_chunk_source const& source,
-                                         cudf::io::text::detail::trie const& trie,
-                                         scan_tile_state<multistate>& tile_multistates,
-                                         scan_tile_state<int64_t>& tile_offsets,
-                                         device_span<int64_t> output_buffer,
-                                         rmm::cuda_stream_view stream,
-                                         std::vector<rmm::cuda_stream_view> const& streams)
+template <typename T>
+class output_builder {
+ public:
+  using size_type = typename rmm::device_uvector<T>::size_type;
+
+  output_builder(size_type max_write_size,
+                 rmm::cuda_stream_view stream,
+                 rmm::mr::device_memory_resource* mr)
+    : _size{0}, _max_write_size{max_write_size}
+  {
+    assert(max_write_size > 0);
+    _chunks.emplace_back(0, stream, mr);
+    _chunks.back().reserve(max_write_size * 2, stream);
+  }
+
+  output_builder(output_builder&&)      = delete;
+  output_builder(const output_builder&) = delete;
+  output_builder& operator=(output_builder&&) = delete;
+  output_builder& operator=(const output_builder&) = delete;
+
+  [[nodiscard]] split_device_span<T> next_output(rmm::cuda_stream_view stream)
+  {
+    auto head_it   = _chunks.end() - (_chunks.size() > 1 && _chunks.back().is_empty() ? 2 : 1);
+    auto head_span = get_free_span(*head_it);
+    if (head_span.size() >= _max_write_size) { return split_device_span<T>{head_span}; }
+    if (head_it == _chunks.end() - 1) {
+      // insert a new vector of double size
+      auto const next_chunk_size = 2 * _chunks.back().capacity();
```

Comment:
Should we reduce the growth rate, given that allocation speed is not really a limiting factor in our context? I'd love to see how different exponents (maybe even 1?) impact the benchmarks.

Reply:
What do you think of https://github.com/rapidsai/cudf/pull/11587/commits/288f2c6f24f45d2f4b9f2c32dc75c93f22b2ebba ("limit overallocation + formatting" in https://github.com/rapidsai/cudf/pull/11587, in case a rebase messes up refs)? Should I pull it in here?

Reply:
I like it! I'm okay with keeping it in the next PR.
Nice to see that we're thinking along the same lines.